streams-tests/jvm/src/test/scala/zio/stream/SinkSpec.scala

```diff
@@ -17,29 +17,791 @@ class SinkSpec(implicit ee: org.specs2.concurrent.ExecutionEnv)
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def mapRemainderInitError = {
+    val sink = initErrorSink.mapRemainder(_.toLong)
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def mapRemainderStepError = {
+    val sink = stepErrorSink.mapRemainder(_.toLong)
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def mapRemainderExtractError = {
+    val sink = extractErrorSink.mapRemainder(_.toLong)
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def optionalHappyPath = {
+    val sink = ZSink.identity[Int].optional
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Some(1)))
+  }
+
+  private def optionalInitError = {
+    val sink = initErrorSink.optional
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== None))
+  }
+
+  private def optionalStepError = {
+    val sink = stepErrorSink.optional
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== None))
+  }
+
+  private def optionalExtractError = {
+    val sink = extractErrorSink.optional
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== None))
+  }
+
+  private def orElseLeft = {
+    val sink = ZSink.identity[Int] orElse ZSink.fail("Ouch")
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Left(1)))
+  }
+
+  private def orElseRight = {
+    val sink = ZSink.fail("Ouch") orElse ZSink.succeedLazy("Hello")
+    unsafeRun(sinkIteration(sink, "whatever").map(_ must_=== Right("Hello")))
+  }
+
+  private def orElseInitErrorLeft = {
+    val sink = initErrorSink orElse ZSink.succeedLazy("Hello")
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Right("Hello")))
+  }
+
+  private def orElseInitErrorRight = {
+    val sink = ZSink.identity[Int] orElse initErrorSink
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Left(1)))
+  }
+
+  private def orElseInitErrorBoth = {
+    val sink = initErrorSink orElse initErrorSink
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def orElseStepErrorLeft = {
+    val sink = stepErrorSink orElse ZSink.succeedLazy("Hello")
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Right("Hello")))
+  }
+
+  private def orElseStepErrorRight = {
+    val sink = ZSink.identity[Int] orElse stepErrorSink
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Left(1)))
+  }
+
+  private def orElseStepErrorBoth = {
+    val sink = stepErrorSink orElse stepErrorSink
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def orElseExtractErrorLeft = {
+    val sink = extractErrorSink orElse ZSink.succeedLazy("Hello")
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Right("Hello")))
+  }
+
+  private def orElseExtractErrorRight = {
+    val sink = ZSink.identity[Int] orElse extractErrorSink
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Left(1)))
+  }
+
+  private def orElseExtractErrorBoth = {
```

Comment:
`orElse` could probably benefit from some more tests around leftover handling and so forth, if you have the cycles for that here